### PR TITLE
[GraphQL] Convert monolithic query into a pattern of small queries

### PR DESCRIPTION
**Pull request by ncksllvn**
## Description
This PR conducts an experiment to find out what the GraphQL performance looks like when each node fragment (page type) has its own self-contained `nodeQuery`, instead of combining all of them into the notorious million-line monolithic GraphQL query that we see sometimes taking close to 20 minutes.

Note that these changes are backwards-compatible, meaning that the preview server, the original monolithic GraphQL query, transformers, etc. all continue to function the same way.

Ultimately, what I did here was -

1. Go through `GetAllPages.graphql.js` to find each type of top-level node fragment we are using. 
2. Open each file representing a node fragment. Change its module to export the original plain fragment for backwards-compatibility, and a self-contained `nodeQuery` for this experiment. Each node fragment relied on a set of global paragraph fragments that normally the node fragment has no direct reference to, so I had to also import the paragraph fragments file and add those into the self-contained `nodeQuery` as well.
3. Write a new method onto `build/drupal/api.js` that does the "individualized" version of the GraphQL query, where it executes in parallel all of the `nodeQuery`s.
4. Call this new method from `metalsmith-drupal` when build type isn't prod.

## Testing done
Sometimes the Drupal GraphQL API (both monolithic and individualized) is "hot" and will serve the entire response in seconds. Seriously, sometimes like 3 seconds. This is likely when Drupal's cache is ready and the content unchanged. The case we are worried about though is when the cache is cold, which was kinda tough to replicate. Here's an example though, where it appears to shave off __12 minutes__ . 

The output is kind of verbose, but obviously we can wrap it with the `verbose` build arg. It's useful to see which specific queries are the most expensive. For instance, the `nodeQuery` to get all VA Forms has a lot of nodes (almost 500) and takes almost 3 minutes on its own, which means it is sometimes the bottleneck. We might be able to make this more efficient by splitting it into two queries, with one taking advantage of Drupal's GraphQL cache.

These result are taken from http://jenkins.vfs.va.gov/blue/organizations/jenkins/testing%2Fvets-website/detail/nps-individual-queries/20/pipeline/100. 

The "Multiple queries" ran on the vagovstaging build whereas the monolithic query ran on the vagovprod build. Both use Drupal's prod environment.

<details>

<summary>Multiple queries - finishes at 3 minutes, 2 seconds </summary>

```
[2021-02-08T14:59:44.610Z] Step 3 of 29 start: Get Drupal content
[2021-02-08T14:59:44.610Z] Drupal content unavailable in local cache: /application/.cache/vagovstaging/drupal/pages.json
[2021-02-08T14:59:44.610Z] Attempting to load Drupal content from API at http://internal-dsva-vagov-prod-cms-2000800896.us-gov-west-1.elb.amazonaws.com
[2021-02-08T14:59:44.610Z] Pulling from Drupal via GraphQL...
[2021-02-08T14:59:44.610Z] 2 page nodes loaded from GetNodeHealthServicesListingPages
[2021-02-08T14:59:44.610Z] GetNodeHealthServicesListingPages done: 1.146s
[2021-02-08T14:59:44.610Z] 13 page nodes loaded from GetNodeHealthCareLocalFacilityPages
[2021-02-08T14:59:44.610Z] GetNodeHealthCareLocalFacilityPages done: 1.154s
[2021-02-08T14:59:44.610Z] 158 page nodes loaded from GetNodeEventPages
[2021-02-08T14:59:44.610Z] GetNodeEventPages done: 1.157s
[2021-02-08T14:59:44.610Z] 1 page nodes loaded from GetNodeOffices
[2021-02-08T14:59:44.610Z] GetNodeOffices done: 1.386s
[2021-02-08T14:59:44.610Z] 2 page nodes loaded from GetNodeHealthCareRegionPages
[2021-02-08T14:59:44.610Z] GetNodeHealthCareRegionPages done: 1.481s
[2021-02-08T14:59:44.610Z] 12 page nodes loaded from GetNodeLandingPages
[2021-02-08T14:59:44.611Z] GetNodeLandingPages done: 1.510s
[2021-02-08T14:59:44.611Z] 29 page nodes loaded from GetNodePressReleasePages
[2021-02-08T14:59:44.611Z] GetNodePressReleasePages done: 1.512s
[2021-02-08T14:59:44.611Z] 3 page nodes loaded from GetNodeEventListingPage
[2021-02-08T14:59:44.611Z] GetNodeEventListingPage done: 2.896s
[2021-02-08T14:59:44.611Z] 108 page nodes loaded from GetNodePersonProfiles
[2021-02-08T14:59:44.611Z] GetNodePersonProfiles done: 3.303s
[2021-02-08T14:59:44.611Z] 2 page nodes loaded from GetNodeLeadershipListingPages
[2021-02-08T14:59:44.611Z] GetNodeLeadershipListingPages done: 3.623s
[2021-02-08T14:59:44.611Z] 1 page nodes loaded from GetNodePublicationListingPages
[2021-02-08T14:59:44.611Z] GetNodePublicationListingPages done: 3.704s
[2021-02-08T14:59:44.611Z] 76 page nodes loaded from GetNodeQa
[2021-02-08T14:59:44.611Z] GetNodeQa done: 3.717s
[2021-02-08T14:59:44.611Z] 42 page nodes loaded from GetNodeNewsStoryPages
[2021-02-08T14:59:44.611Z] GetNodeNewsStoryPages done: 3.851s
[2021-02-08T14:59:44.611Z] 274 page nodes loaded from GetNodePages
[2021-02-08T14:59:44.611Z] GetNodePages done: 4.041s
[2021-02-08T14:59:44.611Z] 1 page nodes loaded from GetNodeBasicLandingPage
[2021-02-08T14:59:44.611Z] GetNodeBasicLandingPage done: 4.631s
[2021-02-08T14:59:44.611Z] 2 page nodes loaded from GetNodeLocationsListingPages
[2021-02-08T14:59:44.611Z] GetNodeLocationsListingPages done: 4.846s
[2021-02-08T14:59:44.611Z] 365 page nodes loaded from GetArchivedNodeEventPages
[2021-02-08T14:59:44.611Z] GetArchivedNodeEventPages done: 4.875s
[2021-02-08T14:59:44.611Z] 13 page nodes loaded from GetNodeMultipleQaPages
[2021-02-08T14:59:44.611Z] GetNodeMultipleQaPages done: 4.878s
[2021-02-08T14:59:44.611Z] 0 page nodes loaded from GetNodeMediaListImages
[2021-02-08T14:59:44.611Z] GetNodeMediaListImages done: 5.020s
[2021-02-08T14:59:44.611Z] 2 page nodes loaded from GetNodeStoryListingPages
[2021-02-08T14:59:44.611Z] GetNodeStoryListingPages done: 5.064s
[2021-02-08T14:59:44.611Z] 2 page nodes loaded from GetNodeVamcOperatingStatusAndAlerts
[2021-02-08T14:59:44.611Z] GetNodeVamcOperatingStatusAndAlerts done: 5.130s
[2021-02-08T14:59:44.611Z] 0 page nodes loaded from GetNodeMediaListVideos
[2021-02-08T14:59:44.611Z] GetNodeMediaListVideos done: 5.883s
[2021-02-08T14:59:44.611Z] 10 page nodes loaded from GetNodeSupportResourcesDetailPage
[2021-02-08T14:59:44.611Z] GetNodeSupportResourcesDetailPage done: 5.917s
[2021-02-08T14:59:44.611Z] 0 page nodes loaded from GetCampaignLandingPages
[2021-02-08T14:59:44.611Z] GetCampaignLandingPages done: 5.988s
[2021-02-08T14:59:44.611Z] 6 page nodes loaded from GetNodeStepByStep
[2021-02-08T14:59:44.611Z] GetNodeStepByStep done: 6.005s
[2021-02-08T14:59:44.611Z] 85 page nodes loaded from GetNodeHealthCareRegionDetailPage
[2021-02-08T14:59:44.611Z] GetNodeHealthCareRegionDetailPage done: 6.082s
[2021-02-08T14:59:44.611Z] 0 page nodes loaded from GetNodeChecklist
[2021-02-08T14:59:44.611Z] GetNodeChecklist done: 6.087s
[2021-02-08T14:59:44.611Z] 2 page nodes loaded from GetNodePressReleaseListingPages
[2021-02-08T14:59:44.611Z] GetNodePressReleaseListingPages done: 6.329s
[2021-02-08T15:00:16.599Z] GetNonNodeWebsiteComponents done: 37.508s
[2021-02-08T15:02:52.973Z] 438 page nodes loaded from GetNodeVaForms
[2021-02-08T15:02:52.973Z] GetNodeVaForms done: 3:01.635 (m:ss.mmm)
[2021-02-08T15:02:52.973Z] Finished all queries - 1649 pages
[2021-02-08T15:02:52.973Z] http://internal-dsva-vagov-prod-cms-2000800896.us-gov-west-1.elb.amazonaws.com response time: : 3:01.642 (m:ss.mmm)
[2021-02-08T15:02:52.973Z] Total time to load content from GraphQL: 3:02.001 (m:ss.mmm)
[2021-02-08T15:02:52.973Z] Drupal successfully loaded!
[2021-02-08T15:02:52.973Z] Successfully piped Drupal content into Metalsmith!
[2021-02-08T15:02:52.973Z] Step 3 end [191751ms]: Get Drupal content
```

</details>

<details>

<summary>Monolithic query: 14 minutes, 48 seconds</summary>

```
[2021-02-08T14:59:44.590Z] Step 3 of 29 start: Get Drupal content
[2021-02-08T14:59:44.590Z] Drupal content unavailable in local cache: /application/.cache/vagovprod/drupal/pages.json
[2021-02-08T14:59:44.590Z] Attempting to load Drupal content from API at http://internal-dsva-vagov-prod-cms-2000800896.us-gov-west-1.elb.amazonaws.com
[2021-02-08T15:14:36.188Z] http://internal-dsva-vagov-prod-cms-2000800896.us-gov-west-1.elb.amazonaws.com response time: : 14:48.555 (m:ss.mmm)
[2021-02-08T15:14:36.188Z] Total time to load content from GraphQL: 14:48.902 (m:ss.mmm)
[2021-02-08T15:14:36.188Z] Drupal successfully loaded!
[2021-02-08T15:14:36.188Z] Successfully piped Drupal content into Metalsmith!
[2021-02-08T15:14:36.188Z] Step 3 end [898942ms]: Get Drupal content
```

</details>

Test this locally but pulling the branch and running `yarn build:content --pull-drupal`. 

## Acceptance criteria
- [x] Peak GraphQL response times are drastically reduced
 
## Definition of done
- [x] Events are logged appropriately
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: src/site/stages/build/drupal/metalsmith-drupal.js [hunk 1]
```diff
@@ -159,7 +159,11 @@ async function getContentViaGraphQL(buildOptions) {
 
     console.time(drupalTimer);
 
-    drupalPages = await contentApi.getAllPages();
+    if (buildOptions.buildtype === ENVIRONMENTS.VAGOVPROD) {
+      drupalPages = await contentApi.getAllPages();
+    } else {
+      drupalPages = await contentApi.getAllPagesViaIndividualGraphQlQueries();
+    }
```
Comment (ncksllvn) [inline src/site/stages/build/drupal/metalsmith-drupal.js:166]: This is the entry point

File: src/site/stages/build/drupal/graphql/landingPage.graphql.js [hunk 2]
```diff
@@ -58,3 +66,43 @@ module.exports = `
...
+queryParamToBeChanged.forEach(param => {
+  regString += `${param}|`;
+});
+
+const regex = new RegExp(`${regString}`, 'g');
+const landingPageFragmentModified = landingPageFragment.replace(
+  regex,
+  updateQueryString,
+);
+
```
Comment (ncksllvn) [inline src/site/stages/build/drupal/graphql/landingPage.graphql.js:80]: I do not understand why we have these string manipulations versus just using the correct value in the actual query, but without it I encountered a GraphQL error so from that I discovered how it is used. I assume there is some CMS-internal context/knowledge about this.

File: src/site/stages/build/drupal/graphql/eventPage.graphql.js [hunk 3]
```diff
@@ -77,3 +78,51 @@ module.exports = `
...
+      { field: "status", value: ["1"], enabled: $onlyPublishedContent },
+      { field: "type", value: ["event"] },
+      { field: "field_datetime_range_timezone.end_value", value: ["${firstDayOfMonthUnixStamp}"], operator: SMALLER_THAN_OR_EQUAL },
+    ]
+  }) {
+    entities {
+      ... eventPage
+    }
+  }
+}
```
Comment (ncksllvn) [inline src/site/stages/build/drupal/graphql/eventPage.graphql.js:121]: `GetNodeEventPages` was expensive, so I decided to try making it more efficient by splitting it into two. The second query gets all events that ended before the start of this month, with the hypothesis being that that event data is unlikely to change and therefore a good candidate to be cached. Then `GetNodeEventPages` just gets whatever events are left over. Not certain that it worked yet but seems promising so far.

**Comment by indytechcook** [comment]
@ncksllvn This is awesome!   

I have a question about the limits.  I can see us blowing through these limits by mid year at the latest.  Is there any pagintaion logic?  If you but it in my apologies for not reading the code correctly.

**Comment by indytechcook** [comment]
> I can see us blowing through these limits by mid year at the latest

Which won't really matter if we move to cms-export.

File: src/site/stages/build/drupal/graphql/GetAllPages.graphql.js [hunk 4]
```diff
@@ -61,35 +61,35 @@ const buildQuery = ({ useTomeSync }) => {
...
+  ${locationListingPage.fragment}
+  ${qaPage.fragment}
+  ${faqMultipleQaPage.fragment}
+  ${stepByStepPage.fragment}
+  ${mediaListImages.fragment}
+  ${checklistPage.fragment}
+  ${mediaListVideos.fragment}
+  ${supportResourcesDetailPage.fragment}
+  ${basicLandingPage.fragment}
+  ${nodeCampaignLandingPage.fragment}
```
Comment (ncksllvn) [inline src/site/stages/build/drupal/graphql/GetAllPages.graphql.js:92]: Each module exporting an addition `.fragment` is what makes these changes backwards-compatible...the same pattern is used by the preview server

**Comment by ncksllvn** [comment]
> I have a question about the limits. I can see us blowing through these limits by mid year at the latest. Is there any pagintaion logic? If you but it in my apologies for not reading the code correctly.

Great suggestion. I just pushed some commits that paginated large datasets, which are executed in parallel, and resulting in a pretty amazing performance gain

```
GetNodeQa | 1s | 76 pages
GetNodeChecklist | 2s | 0 pages
GetNodeMediaListImages | 2s | 0 pages
GetNodeMediaListVideos | 2s | 0 pages
GetCampaignLandingPages | 2s | 0 pages
GetNodeLocationsListingPages | 2s | 2 pages
GetNodeMultipleQaPages | 2s | 13 pages
GetNodeStepByStep | 3s | 6 pages
GetNodeBasicLandingPage | 3s | 1 pages
GetNodeSupportResourcesDetailPage | 3s | 10 pages
GetNodeLeadershipListingPages | 3s | 2 pages
GetNodeEventsSlice4 | 1s | 100 pages
GetNodeVamcOperatingStatusAndAlerts | 4s | 2 pages
GetNonNodeWebsiteComponents | 4s | [n/a] pages
GetNodePublicationListingPages | 4s | 1 pages
GetNodeEventsSlice3 | 1s | 100 pages
GetNodeHealthCareRegionDetailPage | 4s | 85 pages
GetNodeEventListingPage | 1s | 3 pages
GetNodeEventsSlice7 | 2s | 0 pages
GetNodeStoryListingPages | 3s | 2 pages
GetNodeNewsStoryPages | 1s | 42 pages
GetNodeEventsSlice2 | 2s | 100 pages
GetNodeEventsSlice1 | 1s | 100 pages
GetNodeEventsSlice6 | 2s | 23 pages
GetNodeEventsSlice5 | 2s | 100 pages
GetNodePressReleaseListingPages | 1s | 2 pages
GetNodePressReleasePages | 1s | 29 pages
GetNodeHealthServicesListingPages | 1s | 2 pages
GetNodeHealthCareLocalFacilityPages | 1s | 13 pages
GetNodeOffices | 1s | 1 pages
GetNodeHealthCareRegionPages | 1s | 2 pages
GetNodeVaFormsSlice5 | 2s | 38 pages
GetNodeVaFormsSlice3 | 2s | 100 pages
GetNodeVaFormsSlice2 | 2s | 100 pages
GetNodeLandingPages | 2s | 12 pages
GetNodeVaFormsSlice1 | 2s | 100 pages
GetNodeVaFormsSlice4 | 2s | 100 pages
GetNodePagesSlice4 | 2s | 0 pages
GetNodePersonProfileSlice3 | 10s | 8 pages
GetNodePagesSlice2 | 14s | 100 pages
GetNodePagesSlice1 | 14s | 100 pages
GetNodePagesSlice3 | 14s | 74 pages
GetNodePersonProfileSlice1 | 19s | 50 pages
GetNodePersonProfileSlice2 | 19s | 50 pages
```

Bear in mind that queries upwards of a couple seconds were _not_ cached by Drupal, so this is just a direct hit.

**Comment by ncksllvn** [comment]
Referencing Slack thread...https://dsva.slack.com/archives/C01512KN35G/p1612803953030500

With the pagination of large datasets, this is a massive improvement with plenty of avenues for scaling.

image

Bear in mind that any queries with response times over 2 seconds likely have no server-side caching in play.

**Comment by ncksllvn** [comment]
Here is an interesting comparison...I modified the code in this PR to execute the monolithic GetAllPages simultaneously to get these metrics.

image

`438s` - `115s` = 323 seconds, or a ~74% performance gain, although I don't know what caching was in play...

Next, I captured the output of both into JSON files and ran a deep compare. NOT equal...so I wrote a little script to find out more.

```
const fs = require('fs-extra');

const legacy = fs.readJSONSync('legacy-pages.json');
const newb = fs.readJSONSync('new-pages.json');

const oldEntities = legacy.data.nodeQuery.entities;
const newEntities = newb.data.nodeQuery.entities;

const diff = oldEntities.filter(old => {

  if (JSON.stringify(old) === '{}') {
    console.log('empty object, who knows')
    return false;
  }

  return !newEntities.find(n => n.entityBundle === old.entityBundle && n.entityId === old.entityId)
})

console.log(diff)
```
Here's my output 

image


Which is a longwinded way of saying that this produces the same result as the old GetAllPages query but this new query is way way faster and we have many more avenues for scaling.

**Comment by ncksllvn** [comment]
Pointed against the Test CMS instance with 5359 pages and ~12K nodes and with the CMS cache completely flushed beforehand, we saw the GraphQL operation to complete in 8 minutes, 41 seconds on my local machine.

image

This is quick, but the key advantage of this approach is that each of the 184 GraphQL queries will have its own caching layer (and more underneath each), so 9 minutes is the ceiling of what we should ever see (and again, this was on my local machine, not CI.)

On my local just now, the whole operation completed in ~565 seconds (~9.4 minutes) with the GraphQL operation completing in just `33.092s`. This is probably with an effective cache in front of GraphQL. A good amount of the total time was spent doing the 90MB write operation into the `vets-website/.cache` directory, which isn't necessary in CI.

**Review by kelsonic — APPROVED** [review]
LGTM! 🎉 Thank you SO much for doing this!! 🙏 Will this affect how we write our graphql fragments in the future?

**Comment by ncksllvn** [comment]
@kelsonic it will but there is already a pattern set up so that we can follow pretty easily. `src/site/stages/build/drupal/graphql/faqMultipleQa.graphql.js` is a good example of a simple query that also includes some shared paragraph fragments

File: src/site/stages/build/drupal/graphql/paragraph-fragments/staffProfile.paragraph.graphql.js [hunk 5]
```diff
@@ -35,7 +35,7 @@ module.exports = `
                 alt
                 title
                 url
-                derivative(style: _1_1_SQUARE_MEDIUM_THUMBNAIL) {
+                derivative(style: _11SQUAREMEDIUMTHUMBNAIL) {
```
Comment (mpelzsherman) [inline src/site/stages/build/drupal/graphql/paragraph-fragments/staffProfile.paragraph.graphql.js:38]: Curious why this change was needed?

**Review by mpelzsherman — APPROVED** [review]

File: src/site/stages/build/drupal/graphql/paragraph-fragments/staffProfile.paragraph.graphql.js (hunk 5 above)
Comment (ncksllvn) [reply src/site/stages/build/drupal/graphql/paragraph-fragments/staffProfile.paragraph.graphql.js:38]: Using `_1_1_SQUARE_MEDIUM_THUMBNAIL` was reported as a GraphQL error and was relying on  `src/site/utilities/stringHelpers.js` to turn it into `_11SQUAREMEDIUMTHUMBNAIL`. Instead of using the `stringHelper`, I just changed the value, because I can't find any context for why it was like that. Asking the CMS folks if they have any idea, https://dsva.slack.com/archives/CDHBKAL9W/p1612907370100500

**Review by brangi — APPROVED** [review]

File: src/site/stages/build/drupal/graphql/paragraph-fragments/staffProfile.paragraph.graphql.js (hunk 5 above)
Comment (ncksllvn) [reply src/site/stages/build/drupal/graphql/paragraph-fragments/staffProfile.paragraph.graphql.js:38]: Found it...it's safe to just make this change. The string-helpers that did the replacement before is just leftover complexity from a CMS feature flag, https://github.com/department-of-veterans-affairs/vets-website/commit/7b54b495ccddc229a811231aefba897066aa7e35#diff-78d2ba7081dd26ac2a8a81b3bc6c0c80461f62412580c370dff8564362e33ee4R71.

**Comment by kelsonic** [comment]
@department-of-veterans-affairs/frontend-review-group could we get a PR review when you have time please? 🙂

**Review by theodur — APPROVED** [review]
LGTM

File: src/site/stages/build/drupal/menus.js [hunk 6]
```diff
@@ -265,10 +265,9 @@ function makeSection(hostUrl, hub, arrayDepth, promo, pages) {
  * @return {Array} headerData - Menu information formatted for the megaMenu React widget.
  */
 function formatHeaderData(buildOptions, contentData) {
-  if (!contentData?.data) {
+  if (!contentData?.data?.menuLinkContentQuery?.entities) {
     // eslint-disable-next-line no-console
-    console.warn('formatHeaderData has no data');
-    return null;
+    throw new Error('formatHeaderData has no data');
```
Comment (ncksllvn) [inline src/site/stages/build/drupal/menus.js:270]: I made this check stronger because I just can't imagine a case where it is acceptable for a build to report success when the MegaMenu is actually failing to build.

File: src/site/stages/build/drupal/graphql/paragraph-fragments/staffProfile.paragraph.graphql.js (hunk 5 above)
Comment (ncksllvn) [reply src/site/stages/build/drupal/graphql/paragraph-fragments/staffProfile.paragraph.graphql.js:37]: I wrote a ticket to remove this old code, https://github.com/department-of-veterans-affairs/va.gov-team/issues/19722
